crates/sargon/src/keys_collector/collector/derivation_purpose.rs

```diff
@@ -0,0 +1,91 @@
+use crate::prelude::*;
+
+/// The purpose that initiated an interaction with the host to derive keys.
+/// The orchestrator behind this operation is the `KeysCollector`.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum DerivationPurpose {
+    /// When the create account flow, initiates keys collection
+    /// for account VECIs
+    CreatingNewAccount,
+
+    /// When the create persona flow, initiates keys collection
+    /// for identity VECIs
+    CreatingNewPersona,
+
+    /// When applying a security shield to an account, initiates keys collection
+    /// for account MFA
+    SecurifyingAccount,
+
+    /// When applying a security shield to a persona, initiates keys collection
+    /// for identity MFA
+    SecurifyingPersona,
+
+    /// When adding a new factor source, initiates keys collection
+    /// for collecting various factor instances.
+    PreDerivingKeys,
+}
+
+impl DerivationPurpose {
+    pub fn new_for_creating(entity_kind: CAP26EntityKind) -> Self {
+        match entity_kind {
+            CAP26EntityKind::Account => Self::CreatingNewAccount,
+            CAP26EntityKind::Identity => Self::CreatingNewPersona,
+        }
+    }
+
+    pub fn new_for_securifying(entity_kind: CAP26EntityKind) -> Self {
+        match entity_kind {
+            CAP26EntityKind::Account => Self::SecurifyingAccount,
+            CAP26EntityKind::Identity => Self::SecurifyingPersona,
+        }
+    }
+
+    pub fn new_for_pre_derivation() -> Self {
```

Comment:
this function has a strange name IMO. I think it should be called `cache_filling` or `pre_derivation_for_cache` or simply just `pre_deriving_keys` (not a super necessary function)